src/controller.cpp

```diff
@@ -0,0 +1,1255 @@
+                         global_plan_tf_[controller_state_ptr->current_global_plan_index + 1],
+                         pose_projection_ahead, distance2_to_line_ahead, distance_to_end_line_ahead);
+    distToSegmentSquared(current_tf2, global_plan_tf_[controller_state_ptr->current_global_plan_index - 1],
+                         global_plan_tf_[controller_state_ptr->current_global_plan_index],
+                         pose_projection_behind, distance2_to_line_behind, distance_to_end_line_behind);
+
+    if (distance2_to_line_ahead < distance2_to_line_behind)
+    {
+      current_goal_local = pose_projection_ahead;
+      distance_to_goal_ =
+          distance_to_goal_vector_[controller_state_ptr->current_global_plan_index + 1] + distance_to_end_line_ahead;
+      controller_state_ptr->last_visited_pose_index = controller_state_ptr->current_global_plan_index;
+    }
+    else
+    {
+      current_goal_local = pose_projection_behind;
+      distance_to_goal_ =
+          distance_to_goal_vector_[controller_state_ptr->current_global_plan_index] + distance_to_end_line_behind;
+      controller_state_ptr->last_visited_pose_index = controller_state_ptr->current_global_plan_index - 1;
+    }
+    path_pose_idx = controller_state_ptr->current_global_plan_index;
+  }
+  return current_goal_local;
+}
+
+geometry_msgs::Twist Controller::update(const double target_x_vel,
+                                        const double target_end_x_vel,
+                                        const geometry_msgs::Transform current_tf,
+                                        const geometry_msgs::Twist odom_twist,
+                                        const ros::Duration dt,
+                                        double* eda, double* progress, path_tracking_pid::PidDebug* pid_debug)
+{
+  double current_x_vel = controller_state_.current_x_vel;
+  const double current_yaw_vel = controller_state_.current_yaw_vel;
+
+  // Compute location of the point to be controlled
+  double theda_rp = tf2::getYaw(current_tf.rotation);
+  tf2::Vector3 current_with_carrot_origin;
+  current_with_carrot_origin.setX(current_tf.translation.x + l_ * cos(theda_rp));
+  current_with_carrot_origin.setY(current_tf.translation.y + l_ * sin(theda_rp));
+  current_with_carrot_origin.setZ(0);
+
+  current_with_carrot_.setOrigin(current_with_carrot_origin);
+  tf2::Quaternion cur_rot(current_tf.rotation.x, current_tf.rotation.y, current_tf.rotation.z, current_tf.rotation.w);
+  current_with_carrot_.setRotation(cur_rot);
+
+  size_t path_pose_idx;
+  if (track_base_link_enabled_)
+  {
+    // Find closes robot position to path and then project carrot on goal
+    current_pos_on_plan_ = current_goal_ = findPositionOnPlan(current_tf, &controller_state_, path_pose_idx);
+    // To track the base link the goal is then transform to the control point goal
+    double theda_rp = tf2::getYaw(current_goal_.getRotation());
```

Comment:
This value was already calculated above. Does that need to happen again? And only for this `if` scope?